library/src/main/java/io/ona/kujaku/services/TrackingService.java

```diff
@@ -0,0 +1,957 @@
+import android.app.NotificationManager;
+import android.app.PendingIntent;
+import android.app.Service;
+import android.content.Context;
+import android.content.Intent;
+
+import android.content.ServiceConnection;
+import android.graphics.Color;
+
+import android.location.Location;
+import android.location.LocationListener;
+import android.location.LocationManager;
+import android.location.LocationProvider;
+import android.os.Binder;
+import android.os.Build;
+import android.os.Bundle;
+import android.os.Handler;
+import android.os.IBinder;
+import android.os.Looper;
+import android.os.PowerManager;
+import android.support.annotation.NonNull;
+import android.support.v4.app.NotificationCompat;
+import android.support.v4.app.TaskStackBuilder;
+import android.util.Log;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+import io.ona.kujaku.R;
+import io.ona.kujaku.helpers.storage.TrackingStorage;
+import io.ona.kujaku.listeners.TrackingServiceListener;
+import io.ona.kujaku.services.options.TrackingServiceOptions;
+import io.ona.kujaku.services.options.TrackingServiceSaveBatteryOptions;
+
+
+/**
+ * Tracking Service used in Foreground to avoid any memory cleaning from Android
+ * /!\ The application need to be set on Mode "No Battery optimization" in any case
+ *
+ * Created by Emmanuel Otin - eo@novel-t.ch 03/07/19.
+ */
+public class TrackingService extends Service {
+    private final static String TAG = TrackingService.class.getSimpleName();
+    private final static String PARTIAL_WAKE_LOCK_TAG = "TrackingService:PartialWakeLock";
+
+    private final static String ACTIVITY_EXTRA_NAME = "launch_activity_class";
+    private final static String OPTIONS_EXTRA_NAME = "tracking_service_options";
+
+    // Wait time in milli sec to wait for the service thread to exit
+    private final static long WAIT_TIME_SERVICE_THREAD = 400;
+
+    private static volatile CountDownLatch serviceThreadRunningLatch;
+
+    // Service status
+    private static volatile int serviceStatus = TrackingServiceStatus.STOPPED;
+
+    // Location Manager
+    private volatile LocationManager locationManager;
+
+    private volatile Handler gpsHandler;
+    private Handler uiHandler;
+
+    private volatile Location lastRecordedLocation;
+    private volatile Location pendingRecordingLocation;
+    private volatile Location lastBestLocation;
+
+    // Store the recorded locations
+    private List<Location> recordedLocations;
+
+    private volatile Location firstLocationReceived = null ;
+
+    // Tracks Options parameters
+    private TrackingServiceOptions trackingServiceOptions;
+
+    // Use for notification
+    private PendingIntent notificationPendingIntent;
+
+    // Binder given to clients
+    private final IBinder binder = new LocalBinder();
+
+    // Listener to register to some Tracking functions
+    private TrackingServiceListener trackingServiceListener = null;
```

Comment:
I have been testing for memory leaks and you code is pretty good. Listeners attached or implemented in activities or contexts that are going to have a shorter life compared to the event emitter cause memory leaks.

This listener here causes a memory leak of the `PassiveRecordObjectActivity` because there is no way for the activity to nullify the listener when it exits. The service however still holds a reference to the activity through the listener. Bound services fix this by having the unbind

Provide a good way to handle such memory leaks. I can see that `registerTrackingServiceListener` can take in nulls. Provide a way to unregister.

Also fix the memory leak in the sample activity

Reply:
Hi Ephraim, i have fixed this memory leak last week. Could you please tell me how do you test/fin memory leak ? do you use a tool or a library ? Thanks

Reply:
You can generally test for leaks in code using https://github.com/square/leakcanary

Reply:
The leak is fixed

Reply:
ok cool